tap_googleads/custom_query_stream.py

```diff
@@ -0,0 +1,178 @@
+            message = f"The custom GAQL query {self.custom_query['name']} failed. Validate your GAQL query with the Google Ads query validator. https://developers.google.com/google-ads/api/fields/v19/query_validator"
+            raise Exception(message)
+
+        fields = []
+        has_where_clause = False
+        for token in query_object.tokens:
+            if isinstance(token, sqlparse.sql.IdentifierList):
+                fields = [field.strip() for field in token.value.split(",")]
+            if isinstance(token, sqlparse.sql.Where):
+                has_where_clause = True
+
+        date_filter = (
+            f"segments.date >= {self.start_date} and segments.date <= {self.end_date}"
+        )
+        if "segments.date" not in fields:
+            fields.append("segments.date")
+        if has_where_clause:
+            self._query = self._query + f" AND {date_filter}"
+        else:
+            self._query = self._query + f" WHERE {date_filter}"
+
+        google_schema = self.get_fields_metadata(fields)
+
+        for field in fields:
+            node = google_schema[field]
+            google_data_type = node.get("dataType", "")
+            field_value = {
+                "type": [
+                    google_datatype_mapping.get(google_data_type, "string"),
+                    "null",
+                ]
+            }
+
+            if google_data_type == "DATE" and field in DATE_TYPES:
+                field_value["format"] = "date"
+
+            if google_data_type == "ENUM":
+                field_value = {
+                    "type": "string",
+                    "enum": list(node.get("enumValues", [])),
+                }
+
+            if node.get("isRepeated", False):
+                field_value = {"type": ["null", "array"], "items": field_value}
+
+            # GAQL fields look like metrics.cost_micros and response looks like
+            # {'metrics': {'costMicros': 1000000}} which gets converted to metrics__costMicros
+            field_name = "__".join([humps.camelize(i) for i in field.split(".")])
+            local_json_schema["properties"][field_name] = field_value
+        # This is always present in the response
+        local_json_schema["properties"]["customer_id"] = {"type": ["string", "null"]}
+        return local_json_schema
+
+    def _cast_value(self, key: str, value: Any) -> Any:
+        # Some values, notably campaign__id, are returned as strings but the field
+        # data type from the API is integer. This function casts the value to the correct type.
+        if key in self.schema["properties"]:
+            if self.schema["properties"][key]["type"][0] == "integer":
+                return int(value)
+        return value
+
+    def post_process(  # noqa: PLR6301
+        self,
+        row,
+        context=None,
+    ) -> dict | None:
+        flattened_row = flatten_record(
+            record=row,
+            flattened_schema=self.schema,
+            max_level=2,
+        )
```

Comment:
How does this play with the flattening configuration already established in `TapGoogleAds.setup_mapper`?